### PR TITLE
Fix Pickle support for DataFrame and Series

### LINKED ISSUE #672: Modin objects are not able to be pickled
```
File ".../site-packages/cache_memoize/__init__.py", line 116, in inner
    result = cache.get(cache_key, MARKER)
  File ".../site-packages/django_redis/cache.py", line 32, in _decorator
    return method(self, *args, **kwargs)
  File ".../site-packages/django_redis/cache.py", line 81, in get
    client=client)
  File ".../site-packages/django_redis/client/default.py", line 210, in get
    return self.decode(value)
  File ".../site-packages/django_redis/client/default.py", line 318, in decode
    value = self._serializer.loads(value)
  File ".../site-packages/django_redis/serializers/pickle.py", line 35, in loads
    return pickle.loads(value)
  File ".../site-packages/modin/pandas/base.py", line 3337, in __getattribute__
    if self.empty and is_callable:
  File ".../site-packages/modin/pandas/base.py", line 3343, in __getattribute__
    return object.__getattribute__(self, item)
  File ".../site-packages/modin/pandas/dataframe.py", line 211, in empty
    return len(self.columns) == 0 or len(self.index) == 0

  File ".../site-packages/modin/pandas/dataframe.py", line 131, in _get_columns
    return self._query_compiler.columns
  File ".../site-packages/modin/pandas/dataframe.py", line 1898, in __getattr__
    if key in self.columns:
  File ".../site-packages/modin/pandas/base.py", line 3343, in __getattribute__
    return object.__getattribute__(self, item)

#  ... So many times!

  File ".../site-packages/modin/pandas/dataframe.py", line 131, in _get_columns
    return self._query_compiler.columns
  File ".../site-packages/modin/pandas/base.py", line 3343, in __getattribute__
    return object.__getattribute__(self, item)
RecursionError: maximum recursion depth exceeded while calling a Python object
```
 I use django-cache to cache the result of modin df, When django want to fetch the cache,It hit this issue.

### PR DESCRIPTION
**Pull request by vnlitvinov**
## What do these changes do?
* Fix Pickle support for Modin `DataFrame` and `Series` by implementing `__reduce__` protocol (ref: [Python docs](https://docs.python.org/3/library/pickle.html#object.__reduce__))
* Add new protected method `_materialize` to Modin Frame which makes sure all deferred functions for each partition are executed
* Add configuration variable `PersistentPickle` that controls how objects are pickled (backed by `MODIN_PERSISTENT_PICKLE` environment variable): in light mode we store all internals (like partitions), which should be faster for e.g. huge distributed DataFrames (we'll store just arrays of `Ray.ObjectIDs` instead of full data), in full mode it is possible to save pickled objects to disk and restore them in other Modin session but it is slower because we pull in all the data to the driver process
* Extract the construction of the set of default attributes to a global variable - slightly speeds up `__getattribute__` _(unrelated to the PR but small enough so I hope to make it through)_

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #672 (and potentially some more)
- [x] tests added and passing

### CODE REVIEW

File: modin/pandas/series.py [hunk 1]
```diff
@@ -1734,6 +1734,20 @@ def _getitem(self, key):
...
+    @classmethod
+    def _inflate_light(cls, query_compiler, name):
+        return cls(query_compiler=query_compiler, name=name)
+
+    @classmethod
+    def _inflate_full(cls, pandas_series):
+        return cls(data=pandas_series)
+
+    def __reduce__(self):
+        self._query_compiler._modin_frame._materialize()
```
Comment (devin-petersohn) [inline modin/pandas/series.py:1746]: I think this leaks too much into the API layer.

**Review by devin-petersohn — COMMENTED** [review]
Can we avoid accessing deep internals from the API layer?

File: modin/pandas/series.py (hunk 1 above)
Comment (vnlitvinov) [reply modin/pandas/series.py]: I've renamed the method to `.finalize()` and added the same method to query compiler level, so API level does not directly access the frame level anymore.

File: modin/backends/pandas/query_compiler.py [hunk 2]
```diff
@@ -201,6 +201,9 @@ def default_to_pandas(self, pandas_op, *args, **kwargs):
         else:
             return result
 
+    def finalize(self):
+        self._modin_frame.finalize()
```
Comment (anmyachev) [inline modin/backends/pandas/query_compiler.py:205]: What about other backends? (`pyarrow`, `omnisci`) Should we override for these?

Comment (vnlitvinov) [reply modin/backends/pandas/query_compiler.py:205]: We should override it for every backend which defers operations, so most likely yes.

File: modin/pandas/dataframe.py [hunk 3]
```diff
@@ -2416,6 +2413,27 @@ def _getitem(self, key):
         else:
             return self._getitem_column(key)
 
+    # persistance support
```
Comment (anmyachev) [inline modin/pandas/dataframe.py:2416]: Maybe this comment should be above `def _inflate_full` function, not `def _inflate_light`?

Comment (vnlitvinov) [reply modin/pandas/dataframe.py:2416]: this comment denotes the place where persistance support methods start... I should probably make it clearer.

Comment (vnlitvinov) [reply modin/pandas/dataframe.py]: I tried to make it clearer by rewording the comments.

File: modin/backends/pandas/query_compiler.py (hunk 2 above)
Comment (vnlitvinov) [reply modin/backends/pandas/query_compiler.py:205]: Do note that for OmniSci it is already overridden (with `raise NotImplementedError()` stub, cc @ienkovich), and `pyarrow` inherits the one from Pandas so it doesn't seem to be necessary for now.

**Review by devin-petersohn — APPROVED** [review]
@vnlitvinov LGTM! Thanks

**Comment by vnlitvinov** [comment]
@devin-petersohn CI is green but code coverage (which is weird at best as of late anyway).

**Comment by devin-petersohn** [comment]
Awaiting approval from @modin-project/modin-omnisci since there were code changes there as well.

**Review by ienkovich — APPROVED** [review]
LGTM!

**Pull request merged**